internal/log_analysis/alerts_api/table/list.go

```diff
@@ -26,84 +26,254 @@ import (
 	jsoniter "github.com/json-iterator/go"
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
+
+	"github.com/panther-labs/panther/api/lambda/alerts/models"
 )
 
-func (table *AlertsTable) ListByRule(ruleID string, exclusiveStartKey *string, pageSize *int) (
+// ListByRule - lists all alerts belonging to a specific ruleID
+func (table *AlertsTable) ListByRule(input *models.ListAlertsInput) (
 	summaries []*AlertItem, lastEvaluatedKey *string, err error) {
 
-	return table.list(RuleIDKey, ruleID, exclusiveStartKey, pageSize)
+	return table.list(RuleIDKey, *input.RuleID, input)
 }
 
-func (table *AlertsTable) ListAll(exclusiveStartKey *string, pageSize *int) (
+// ListAll - lists all alerts (default sort by creation time)
+func (table *AlertsTable) ListAll(input *models.ListAlertsInput) (
 	summaries []*AlertItem, lastEvaluatedKey *string, err error) {
 
-	return table.list(TimePartitionKey, TimePartitionValue, exclusiveStartKey, pageSize)
+	return table.list(TimePartitionKey, TimePartitionValue, input)
 }
 
-// list returns a page of alerts ordered by creationTime, last evaluated key, any error
-func (table *AlertsTable) list(ddbKey, ddbValue string, exclusiveStartKey *string, pageSize *int) (
-	summaries []*AlertItem, lastEvaluatedKey *string, err error) {
+// getIndex - gets the primary index to query
+//
+// If a `RuleID` is present, then create the index based on this field.
+// Otherwise, use the default time partition index
+func (table *AlertsTable) getIndex(input *models.ListAlertsInput) (index *string) {
+	if input.RuleID != nil {
+		index = aws.String(table.RuleIDCreationTimeIndexName)
+		return index
+	}
+	index = aws.String(table.TimePartitionCreationTimeIndexName)
+	return index
+}
 
-	// pick index
-	var index string
-	if ddbKey == RuleIDKey {
-		index = table.RuleIDCreationTimeIndexName
-	} else if ddbKey == TimePartitionKey {
-		index = table.TimePartitionCreationTimeIndexName
-	} else {
-		return nil, nil, errors.New("unknown key" + ddbKey)
+// getKeyBuilder - gets the appropriate key builder
+func (table *AlertsTable) getKeyBuilder(input *models.ListAlertsInput) (keyBuilder expression.KeyBuilder) {
+	if input.RuleID != nil {
+		keyBuilder = expression.Key(RuleIDKey)
+		return keyBuilder
+	}
+	keyBuilder = expression.Key(TimePartitionKey)
+	return keyBuilder
+}
+
+// getKeyCondition - gets the key condition for a query
+func (table *AlertsTable) getKeyCondition(keyBuilder *expression.KeyBuilder,
+	input *models.ListAlertsInput) (keyCondition expression.KeyConditionBuilder) {
```

Comment:
Nit: You don't need a named return

```suggestion
	input *models.ListAlertsInput) expression.KeyConditionBuilder
```

Then you can simply `return keyBuilder.Equal(...)` in each case